cpp/inc/bond/core/maybe.h

```diff
@@ -14,140 +22,541 @@ namespace bond
+    ///
+    /// @since 8.0.0
+    friend bool operator!=(const maybe_common& lhs, const maybe_common& rhs)
     {
-        _nothing = true;
+        return !(lhs == rhs);
     }
 
-    /// @brief Reference to the value
-    /// @throw NothingException if the object contains nothing
-    T& value()
+    /// @brief Compares a maybe and a value for equality.
+    ///
+    /// @return false if the maybe holds nothing; otherwise, calls
+    /// operator== with the maybe's value and the provided value.
+    ///
+    /// @since 8.0.0
+    friend bool operator==(const maybe_common& lhs, const T& rhs)
     {
-        if (_nothing)
-            NothingException();
+        return lhs._value == rhs;
+    }
 
-        return _value;
+    /// @brief Compares a maybe and a value for inequality.
+    ///
+    /// See operator==(const maybe_common&,const T&) for details about how
+    /// maybes holding nothing are handled.
+    ///
+    /// @since 8.0.0
+    friend bool operator!=(const maybe_common& lhs, const T& rhs)
+    {
+        return !(lhs == rhs);
     }
 
-    /// @brief Constant reference to the value
-    /// @throw NothingException if the object contains nothing
-    const T& value() const
+    /// @brief Compares a value and a maybe for equality.
+    ///
+    /// @return false if the maybe holds nothing; otherwise, calls
+    /// operator== with the provided value and maybe's value.
+    ///
+    /// @since 8.0.0
+    friend bool operator==(const T& lhs, const maybe_common& rhs)
     {
-        if (_nothing)
-            NothingException();
+        return lhs == rhs._value;
+    }
 
-        return _value;
+    /// @brief Compares and a value and a maybe for inequality.
+    ///
+    /// See operator==(const T&,const maybe_common&) for details about how
+    /// maybes holding nothing are handled.
+    ///
+    /// @since 8.0.0
+    friend bool operator!=(const T& lhs, const maybe_common& rhs)
+    {
+        return !(lhs == rhs);
     }
 
-    /// @brief Compare for equality
-    bool operator==(const maybe& that) const
+protected:
+    maybe_common() = default;
+
+    #if defined(_MSC_VER) && _MSC_VER < 1900
+    // Using = default with MSVC 2013 on this function causes the compiler
+    // to make extra copies, which makes this type fail when it's holding a
+    // move-only type.
+    //
+    // Manually implementing this function works just fine, however.
+    maybe_common(const maybe_common& that)
+        : _value(that._value)
+    { }
+    #else
+    maybe_common(const maybe_common&) = default;
+    #endif
+
+    template <typename... Args>
+    explicit maybe_common(const T& value, Args&&... args)
     {
-        return _nothing == that._nothing
-            && (_nothing || _value == that._value);
+        _value.emplace(value, std::forward<Args>(args)...);
     }
 
-    bool operator!=(const maybe& that) const
+    template <typename... Args>
+    explicit maybe_common(T&& value, Args&&... args)
+        : _value()
```

Comment:
Can you check all the redundant constructor calls? :)

Reply:
Removed the handful that were left.